crates/miden-proving-service-client/src/remote_prover/batch_prover.rs

```diff
@@ -0,0 +1,105 @@
+use alloc::{
+    boxed::Box,
+    string::{String, ToString},
+    sync::Arc,
+};
+
+use miden_objects::batch::{ProposedBatch, ProvenBatch};
+use miden_tx_batch_prover::errors::BatchProveError;
+use tokio::sync::Mutex;
+
+use super::generated::api_client::ApiClient;
+use crate::RemoteProverError;
+
+// REMOTE BATCH PROVER
+// ================================================================================================
+
+/// A [`RemoteBatchProver`] is a batch prover that sends a proposed batch data to a remote
+/// gRPC server and receives a proven batch.
+///
+/// When compiled for the `wasm32-unknown-unknown` target, it uses the `tonic_web_wasm_client`
+/// transport. Otherwise, it uses the built-in `tonic::transport` for native platforms.
+///
+/// The transport layer connection is established lazily when the first transaction is proven.
+pub struct RemoteBatchProver {
+    #[cfg(target_arch = "wasm32")]
+    client: Arc<Mutex<Option<ApiClient<tonic_web_wasm_client::Client>>>>,
+
+    #[cfg(not(target_arch = "wasm32"))]
+    client: Arc<Mutex<Option<ApiClient<tonic::transport::Channel>>>>,
+
+    endpoint: String,
+}
+
+impl RemoteBatchProver {
+    /// Creates a new [RemoteBatchProver] with the specified gRPC server endpoint. The
+    /// endpoint should be in the format `{protocol}://{hostname}:{port}`.
+    pub fn new(endpoint: &str) -> Self {
+        RemoteBatchProver {
+            endpoint: endpoint.to_string(),
+            client: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Establishes a connection to the remote transaction prover server. The connection is
+    /// maintained for the lifetime of the prover. If the connection is already established, this
+    /// method does nothing.
+    async fn connect(&self) -> Result<(), RemoteProverError> {
+        let mut client = self.client.lock().await;
+        if client.is_some() {
+            return Ok(());
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        let new_client = {
+            let web_client = tonic_web_wasm_client::Client::new(self.endpoint.clone());
+            ApiClient::new(web_client)
+        };
+
+        #[cfg(not(target_arch = "wasm32"))]
+        let new_client = {
+            ApiClient::connect(self.endpoint.clone())
+                .await
+                .map_err(|_| RemoteProverError::ConnectionFailed(self.endpoint.to_string()))?
+        };
+
+        *client = Some(new_client);
+
+        Ok(())
+    }
+}
+
+impl RemoteBatchProver {
+    pub async fn prove(
+        &self,
+        proposed_batch: ProposedBatch,
+    ) -> Result<ProvenBatch, BatchProveError> {
```

Comment:
Does this have to return a `BatchProveError` or could it also return `RemoteProverError`? Then we could have the `Other` variant in `RemoteProveError` instead of `BatchProveError`.
For now, we only need the other variant because of the connection errors, and this seems more related to `RemoteProverError` than the `BatchProveError`. Wdyt?